### PR TITLE
Use external JRuby optional

### PR DESCRIPTION
**Pull request by dmikurube**
Finally! After refactorings and updates (#1332, #1337. #1333, #1334), we remove JRuby out of Embulk executable binary, so that
* Users can specify their own JRuby JAR by an Embulk System Property `jruby`, and
* JRuby is loaded in a sub `ClassLoader` not to conflict with Java-based plugins.

Example command lines are:
* `embulk -X jruby=file:///path/to/jruby-9.1.15.0-complete.jar run test.yml`
* `embulk -X jruby=file:///path/to/jruby-9.1.15.0-complete.jar gem install embulk-input-s3`

### CODE REVIEW

**Comment by dmikurube** [comment]
Ah, I found that it's not working for `embulk gem/bundle` commands after some changes. Let me try fixing it.

**Comment by dmikurube** [comment]
Fixed the issue above ^ in 509411cc0a987fb72bac7044f594a26a29953c1f.

We may want also to refactor the JRuby-related parts... in another PR sooner.

**Review by trung-huynh — APPROVED** [review]
LGTM 👍

File: embulk-core/src/main/java/org/embulk/jruby/LazyScriptingContainerDelegate.java [hunk 1]
```diff
@@ -18,6 +27,101 @@ public LazyScriptingContainerDelegate(
...
+        if (jrubyProperty != null && !jrubyProperty.isEmpty()) {
+            // File.pathSeparator is not available here because the property "jruby" expects a URL-like style:
+            // "file:" and "mvn:". It now supports only "file:", though.
+            //
+            // Semicolons basically do not appear in URLs without quoting except for the "optional fields and values"
+            // case described in RFC 1738. We don't need to take care of the "optional fields and values" case here.
+            // https://tools.ietf.org/html/rfc1738
+            for (final String jarLocator : jrubyProperty.split("\\;")) {
+                if (jarLocator.startsWith("file:")) {
+                    // TODO: Validate the path more.
```
Comment (shroman) [inline embulk-core/src/main/java/org/embulk/jruby/LazyScriptingContainerDelegate.java:71]: Is this TODO a leftover on purpose?

Comment (dmikurube) [reply embulk-core/src/main/java/org/embulk/jruby/LazyScriptingContainerDelegate.java:71]: Ah, not on a very specific purpose, but it was just a possible future improvement for providing a better error message / log for a case that a user configures `jruby` wrongly.

It is now setting user's configuration directly to `URLClassLoader`, so a user may see some cryptic error messages when the configuration is wrong. I thought it's worth validating user's configuration and providing a better error message, but wanted to think about that later...

File: embulk-core/src/main/java/org/embulk/EmbulkRunner.java [hunk 2]
```diff
@@ -373,11 +374,8 @@ private String runLiquid(
...
         // TODO: Check if it is required to process JRuby options.
-        // Not |ScriptingContainerDelegate.LocalContextScope.SINGLETON| to narrow down considerations.
-        final ScriptingContainerDelegate localJRubyContainer = ScriptingContainerDelegateImpl.create(
-                EmbulkRunner.class.getClassLoader(),
-                ScriptingContainerDelegate.LocalContextScope.SINGLETHREAD,
-                ScriptingContainerDelegate.LocalVariableBehavior.PERSISTENT);
+        final ScriptingContainerDelegate localJRubyContainer =
+                LazyScriptingContainerDelegate.withGems(rootLogger, this.embulkSystemProperties);
 
         localJRubyContainer.runScriptlet("require 'liquid'");
```
Comment (shroman) [inline embulk-core/src/main/java/org/embulk/EmbulkRunner.java:380]: Looks like `localJRubyContainer` can be `null`.

File: embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        if (parentUrl != null) {
+            return parentUrl;
+        }
+
+        return this.findResource(name);
+    }
+
+    @Override
+    public Enumeration<URL> getResources(final String name) throws IOException {
+        final Vector<URL> resources = new Vector<URL>();
```
Comment (shroman) [inline embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java:99]: nit: can be `final Vector<URL> resources = new Vector<>();`

File: embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Loads classes and resources of JRuby artifact(s).
+ *
+ * <p>It prioritizes a class found by this class loader over a class loaded by the parent class loader because
+ * this class loader is specific for JRuby. {@code jruby-complete-9.X.Y.Z.jar} contains all libraries required
+ * in JRuby. Some of them conflict with {@code embulk-core}'s top-level dependencies, such as {@code "asm:asm"}
+ * and {@code "org.sonatype.sisu.inject:cglib"} used from Guice.
+ *
+ * <p>JRuby should use JRuby's own dependencies, and Embulk should not be impacted from JRuby's depencencies.
```
Comment (shroman) [inline embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java:18]: nit: typo in `depencencies`.

File: embulk-core/src/main/java/org/embulk/jruby/LazyScriptingContainerDelegate.java (hunk 1 above)
Comment (shroman) [reply embulk-core/src/main/java/org/embulk/jruby/LazyScriptingContainerDelegate.java:71]: Got it. Not necessary to get everything into this PR, but some TODOs alert me :)

File: embulk-core/src/main/java/org/embulk/EmbulkRunner.java (hunk 2 above)
Comment (dmikurube) [reply embulk-core/src/main/java/org/embulk/EmbulkRunner.java:386]: Thanks! Added a (very simple) error messaging there also with `gem` and `bundle` in ef7af455559cec2eaf52ed5541bc503de7ae3cf0. We may want to improve those error handling later...

File: embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java (hunk 4 above)
Comment (dmikurube) [reply embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java]: Fixed in ef7af455559cec2eaf52ed5541bc503de7ae3cf0.

File: embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java (hunk 3 above)
Comment (dmikurube) [reply embulk-core/src/main/java/org/embulk/jruby/JRubyClassLoader.java]: Fixed in ef7af455559cec2eaf52ed5541bc503de7ae3cf0.

**Review by shroman — APPROVED** [review]
Thanks for addressing the comments. LGTM.

**Comment by dmikurube** [comment]
Thanks! Merging, and releasing `v0.10.21` soon...

**Pull request merged**